### PR TITLE
Cache buildbot builds on CI server

### PR DESCRIPTION
**Pull request by jacob-hughes**

### CODE REVIEW

**Comment by jacob-hughes** [comment]
bors try

File: .buildbot.sh [hunk 1]
```diff
@@ -8,3 +8,32 @@ set -e
...
+git show -s HEAD > ${TARBALL_TOPDIR}/VERSION
+cd build
+tar jcf ${TARBALL_NAME} `basename ${TARBALL_TOPDIR}`
+chmod 775 ${TARBALL_NAME}
+mv ${TARBALL_NAME} ${SNAP_DIR}
+ln -sf ${SNAP_DIR}/${TARBALL_NAME} ${SNAP_DIR}/${SYMLINK_NAME}
+
+# Remove all but the 10 latest builds
+cd ${SNAP_DIR}
+sh -c "ls -tp | grep -v '/$' | tail -n +11 | xargs -I {} rm -- {}"
```
Comment (ltratt) [inline .buildbot.sh:39]: I suggest we keep fewer builds (with ykrustc we've never made use of this feature): say 3? So it would be `tail -n +4`.

**Comment by jacob-hughes** [comment]
Running the test suite with rustc debug assertions seems to re-order statements in the MIR, something which the diff in `tests/mir-opt/prevent_early_finalization.rs` was not expecting. We weren't running the suite with these assertions before, so I've removed that here:  74e0618

File: .buildbot.sh (hunk 1 above)
Comment (jacob-hughes) [reply .buildbot.sh]: Sure. Fixed 55f3fa3

**Comment by jacob-hughes** [comment]
bors try

**Comment by ltratt** [comment]
If the `try` succeeds, please squash.

**Comment by jacob-hughes** [comment]
I think this assumes that the directory in `opt/` has already been setup on b13 with the correct permissions. Try again now that it has.
bors try

**Comment by ltratt** [comment]
I think I've now fixed the permissions.

bors try

**Comment by jacob-hughes** [comment]
That OOM'd for some reason. I wonder if it was because it was run while another build was running. I'll kick it once more and if not have a proper look tomorrow.

bors try

**Comment by ltratt** [comment]
Looks we're good to go! Please squash.

File: .buildbot.config.toml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+[build]
+docs = false
+extended = true
+tools = ["cargo", "rustdoc"]
+
+[llvm]
+assertions = true           # Turn on assertions in LLVM.
+
+[install]
+prefix = "build/rustgc-stage2-latest"
```
Comment (vext01) [inline .buildbot.config.toml:14]: I often wonder why I used "latest" here. It's like saying something is "final": it's doomed to go stale.

**Comment by jacob-hughes** [comment]
Squashed

**Comment by ltratt** [comment]
@jacob-hughes Any luck with this one?

**Comment by jacob-hughes** [comment]
No, this one is a bit strange. I can't reproduce it locally but I'm trying on b13. It's a new one - it hadn't come up before squashing :(

**Comment by vext01** [comment]
> fatal runtime error: Rust cannot catch foreign exceptions
> rustc exited with signal: 6

**Comment by ltratt** [comment]
Has this perhaps exceeded the buildbot/bors timeout?

**Comment by jacob-hughes** [comment]
Maybe. That usually ends up throwing a timeout error though. I can bump the ulimit value in the `.buildbot.sh` script too, but I'm running out of ideas after that.

**Comment by ltratt** [comment]
Give them both a go and see?!

**Comment by jacob-hughes** [comment]
It doesn't seem to be that. I've managed to replicate this issue locally so am now trying to work out what's going on.

**Comment by vext01** [comment]
Could it be to do with assertions that you've not usually had enabled, but have enabled in the CI config.toml?

**Comment by vext01** [comment]
This line, for example:

```
debug-assertions = true     # Turn on assertions in rustc.
```

**Comment by jacob-hughes** [comment]
I've rebased this against the current master now. :crossed_fingers: it builds.

bors try

**Comment by ltratt** [comment]
OK to merge?

**Comment by jacob-hughes** [comment]
:tada: Yes please!

**Pull request merged**